pydrake doc: gcc and clang disagree on need for lambda capture for `constexpr` vars

From here: https://github.com/RobotLocomotion/drake/pull/9571#issuecomment-426708850

Workarounds:
- Use `constexpr auto& doc`, rather than `auto& doc`
- Use `[&]` in lambda capture
- Add `copts = ["-Wno-unused-lambda-capture"]` for `pybind_py_library`

Action items:
- [x] Use a workaround in relevant sites (searching for `doc_lambda`)
- [x] Find or file upstream bug report:
   - clang: https://bugs.llvm.org/show_bug.cgi?id=39202
   - gcc: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87559

\cc @m-chaturvedi @jamiesnape

From chat with @m-chaturvedi, seems that GCC 7.3.0 on Bionic decided to further deviate from `clang`; with the merge of #9606, we now get a Bionic failure:
https://drake-jenkins.csail.mit.edu/job/linux-bionic-gcc-bazel-experimental/11/console
```
bindings/pydrake/systems/framework_py_semantics.cc:72:37: note: the lambda has no capture-default
   auto bind_common_scalar_types = [m](auto dummy) {
                                     ^
bindings/pydrake/systems/framework_py_semantics.cc:31:19: note: 'constexpr const<unnamed struct>::<unnamed struct>::<unnamed struct>& doc' declared here
   constexpr auto& doc = pydrake_doc.drake.systems;
                   ^~~
bindings/pydrake/systems/framework_py_semantics.cc:310:12: error: 'doc' is not captured
            doc.Parameters.SetFrom.doc);
```

Will submit a chaser PR to use the second workaround :(

--- a/bindings/pydrake/automotive_py.cc
+++ b/bindings/pydrake/automotive_py.cc
@@ -26,7 +26,7 @@ PYBIND11_MODULE(automotive, m) {
   using namespace drake::automotive;
 
   m.doc() = "Bindings for Automotive systems";
-  auto& doc = pydrake_doc.drake.automotive;
+  constexpr auto& doc = pydrake_doc.drake.automotive;
 
   py::module::import("pydrake.systems.framework");
   py::module::import("pydrake.systems.rendering");

--- a/bindings/pydrake/geometry_py.cc
+++ b/bindings/pydrake/geometry_py.cc
@@ -32,7 +32,7 @@ void BindIdentifier(py::module m, const std::string& name) {
 PYBIND11_MODULE(geometry, m) {
   // NOLINTNEXTLINE(build/namespaces): Emulate placement in namespace.
   using namespace drake::geometry;
-  auto& doc = pydrake_doc.drake.geometry;
+  constexpr auto& doc = pydrake_doc.drake.geometry;
 
   py::module::import("pydrake.systems.framework");
   py::class_<SceneGraph<T>, LeafSystem<T>>(m, "SceneGraph", doc.SceneGraph.doc)

--- a/bindings/pydrake/lcm_py.cc
+++ b/bindings/pydrake/lcm_py.cc
@@ -17,7 +17,7 @@ namespace pydrake {
 PYBIND11_MODULE(lcm, m) {
   // NOLINTNEXTLINE(build/namespaces): Emulate placement in namespace.
   using namespace drake::lcm;
-  auto& doc = pydrake_doc.drake.lcm;
+  constexpr auto& doc = pydrake_doc.drake.lcm;
 
   // Use `py::bytes` as a mid-point between C++ LCM (`void* + int` /
   // `vector<uint8_t>`) and Python LCM (`str`).

--- a/bindings/pydrake/math_py.cc
+++ b/bindings/pydrake/math_py.cc
@@ -21,7 +21,7 @@ PYBIND11_MODULE(math, m) {
   using namespace drake::math;
 
   m.doc() = "Bindings for //math.";
-  auto& doc = pydrake_doc.drake.math;
+  constexpr auto& doc = pydrake_doc.drake.math;
 
   py::module::import("pydrake.util.eigen_geometry");
 

--- a/bindings/pydrake/multibody/rigid_body_tree_py.cc
+++ b/bindings/pydrake/multibody/rigid_body_tree_py.cc
@@ -258,7 +258,8 @@ PYBIND11_MODULE(rigid_body_tree, m) {
   // `rigid_body_tree.cc`; if the method is not yet bound, the name has a
   // comment as a placeholder.
   // Methods of type (b) are declared below methods of type (a).
-  auto add_rigid_body_tree_typed_methods = [m, &tree_cls](auto dummy) {
+  // N.B. Capturing `&doc` should not be required; workaround per #9600.
+  auto add_rigid_body_tree_typed_methods = [m, &doc, &tree_cls](auto dummy) {
     // N.B. The header files use `Scalar` as the scalar-type template
     // parameter, but `T` is used here for brevity.
     using T = decltype(dummy);

--- a/bindings/pydrake/symbolic_py.cc
+++ b/bindings/pydrake/symbolic_py.cc
@@ -23,7 +23,7 @@ using std::string;
 PYBIND11_MODULE(_symbolic_py, m) {
   // NOLINTNEXTLINE(build/namespaces): Emulate placement in namespace.
   using namespace drake::symbolic;
-  auto& doc = pydrake_doc.drake.symbolic;
+  constexpr auto& doc = pydrake_doc.drake.symbolic;
 
   // Install NumPy warning filtres.
   // N.B. This may interfere with other code, but until that is a confirmed

--- a/bindings/pydrake/systems/framework_py_semantics.cc
+++ b/bindings/pydrake/systems/framework_py_semantics.cc
@@ -69,7 +69,8 @@ void DefineFrameworkPySemantics(py::module m) {
     .def("CopyFrom", &AbstractValues::CopyFrom,
          doc.AbstractValues.CopyFrom.doc);
 
-  auto bind_common_scalar_types = [m](auto dummy) {
+  // N.B. Capturing `&doc` should not be required; workaround per #9600.
+  auto bind_common_scalar_types = [m, &doc](auto dummy) {
     using T = decltype(dummy);
     DefineTemplateClassWithDefault<Context<T>>(
         m, "Context", GetPyParam<T>(), doc.Context.doc)

--- a/bindings/pydrake/systems/framework_py_values.cc
+++ b/bindings/pydrake/systems/framework_py_values.cc
@@ -26,7 +26,8 @@ void DefineFrameworkPyValues(py::module m) {
   using namespace drake::systems;
   constexpr auto& doc = pydrake_doc.drake.systems;
 
-  auto bind_common_scalar_types = [m](auto dummy) {
+  // N.B. Capturing `&doc` should not be required; workaround per #9600.
+  auto bind_common_scalar_types = [m, &doc](auto dummy) {
     using T = decltype(dummy);
     // Value types.
     DefineTemplateClassWithDefault<VectorBase<T>>(

--- a/bindings/pydrake/systems/primitives_py.cc
+++ b/bindings/pydrake/systems/primitives_py.cc
@@ -36,7 +36,8 @@ PYBIND11_MODULE(primitives, m) {
   constexpr auto& doc = pydrake_doc.drake.systems;
 
   py::module::import("pydrake.systems.framework");
-  auto bind_common_scalar_types = [m](auto dummy) {
+  // N.B. Capturing `&doc` should not be required; workaround per #9600.
+  auto bind_common_scalar_types = [m, &doc](auto dummy) {
     using T = decltype(dummy);
 
     DefineTemplateClassWithDefault<Adder<T>, LeafSystem<T>>(

--- a/bindings/pydrake/trajectories_py.cc
+++ b/bindings/pydrake/trajectories_py.cc
@@ -13,7 +13,7 @@ namespace pydrake {
 PYBIND11_MODULE(trajectories, m) {
   // NOLINTNEXTLINE(build/namespaces): Emulate placement in namespace.
   using namespace drake::trajectories;
-  auto& doc = pydrake_doc.drake.trajectories;
+  constexpr auto& doc = pydrake_doc.drake.trajectories;
 
   using T = double;
 

--- a/tools/skylark/pybind.bzl
+++ b/tools/skylark/pybind.bzl
@@ -62,6 +62,11 @@ def pybind_py_library(
         # This is how you tell Bazel to create a shared library.
         linkshared = 1,
         linkstatic = 1,
+        copts = [
+            # GCC and Clang don't always agree / succeed when inferring storage
+            # duration (#9600). Workaround it for now.
+            "-Wno-unused-lambda-capture",
+        ],
         # Always link to pybind11.
         deps = [
             "@pybind11",

